com.unity.cinemachine/Runtime/Behaviours/CinemachineFreeLook.cs

```diff
@@ -439,40 +439,41 @@ float GetYAxisClosestValue(Vector3 cameraPos, Vector3 up)
                 }
                 dir.x = 0;
 
-                // Sample the spline in a few places, find the 2 closest, and lerp
-                int i0 = 0, i1 = 0;
-                float a0 = 0, a1 = 0;
-                const int NumSamples = 13;
-                float step = 1f / (NumSamples-1);
-                for (int i = 0; i < NumSamples; ++i)
-                {
-                    float a = Vector3.SignedAngle(
-                        dir, GetLocalPositionForCameraFromInput(i * step), Vector3.right);
-                    if (i == 0)
-                        a0 = a1 = a;
-                    else
-                    {
-                        if (Mathf.Abs(a) < Mathf.Abs(a0))
-                        {
-                            a1 = a0;
-                            i1 = i0;
-                            a0 = a;
-                            i0 = i;
-                        }
-                        else if (Mathf.Abs(a) < Mathf.Abs(a1))
-                        {
-                            a1 = a;
-                            i1 = i;
-                        }
-                    }
-                }
-                if (Mathf.Sign(a0) == Mathf.Sign(a1))
-                    return i0 * step;
-                float t = Mathf.Abs(a0) / (Mathf.Abs(a0) + Mathf.Abs(a1));
-                return Mathf.Lerp(i0 * step, i1 * step, t);
+                // We need to find the minimum of the angle of function using steepest descent
+                return SteepestDescent(0f, 1f, dir);
             }
             return m_YAxis.Value; // stay conservative
         }
+        
+        float SteepestDescent(float min, float max, Vector3 desiredDirection)
+        {
+            const int maxIteration = 10;
+            const float epsilon = 0.00005f;
+            var x = (min + max) / 2f; // initial guess - mid point
+            for (var i = 0; i < maxIteration; ++i)
+            {
+                var angle = AngleFunction(x);
+                var slope = SlopeOfAngleFunction(x);
+                if (slope == 0 || Mathf.Abs(angle) < epsilon)
+                    break; // found best
+                x = Mathf.Clamp(x - (angle / slope), min, max); // clamping is needed so we don't overshoot
+            }
+            return x;
+
+            // localFunctions
+            float AngleFunction(float input)
+            {
+                var point = GetLocalPositionForCameraFromInput(input);
+                return Vector3.SignedAngle(desiredDirection, point, Vector3.right);
```

Comment:
Vector3.SignedAngle is noisy when the angle is small - that's why you're getting imperfect results.  Try changing it to UnityVectorExtensions.SignedAngle (also above, in the calling function)